Packs/PaloAltoNetworks_IoT3rdParty/Scripts/SendAllPANWIoTDevicesToCiscoISE/SendAllPANWIoTDevicesToCiscoISE.py

```diff
@@ -0,0 +1,247 @@
+        "id": ep_id,
+        "macAddress": mac,
+        "attributeName": attribute_names,
+        "attributeValue": attribute_values,
+        "using": active_instance
+    })
+    if isError(resp[0]):
+        err_msg = f'Error, failed to update custom attributes for endpoint {id} - {resp[0].get("Contents")}'
+        raise Exception(err_msg)
+
+
+def create_new_ep(mac, attr_map, active_instance):
+    """
+    Create a new endpoint with the given params
+    Param mac: mac address of the endpoint that needs to be created.
+    Param attr_map: a map containing various ise custom attributes.
+    Param active_instance: The primary/active ISE instance.
+    """
+    resp = demisto.executeCommand("cisco-ise-create-endpoint", {
+        "mac_address": mac,
+        "attributes_map": attr_map,
+        "using": active_instance
+    })
+    if isError(resp[0]):
+        err_msg = f'Failed to create new Endpoint {mac} - {resp[0].get("Contents")}'
+        raise Exception(err_msg)
+
+
+def create_or_update_ep(mac, attr_map):
+    """
+    Check if an enpoint exists in ISE, if not create one with the custom attributes
+    otherwise update it. If at any point the connection goes down or we get a 401 -
+    unautherized access we will attempt to get the new active instance.
+    Params mac: Mac adress of the endpoint.
+    attr_map: Custom attributes for the endpoint.
+    """
+
+    global CISCO_ISE_ACTIVE_INSTANCE
+    global GET_EP_ID_CMD
+
+    cmd_mac_syntax_map = {
+        "cisco-ise-get-endpoint-id-by-name": "mac_address",
+        "cisco-ise-get-endpoint-id": "macAddress"
+    }
+
+    # Check if this mac address (endpoint) is present in ISE by attempting to get its ID
+    resp = demisto.executeCommand(GET_EP_ID_CMD, {
+        cmd_mac_syntax_map[GET_EP_ID_CMD]: mac,
+        "using": CISCO_ISE_ACTIVE_INSTANCE
+    })
+
+    if isError(resp[0]):
+        err_msg = extract_ise_api_error(resp[0].get("Contents"))
+
+        # 404 Not Found or empty results, we need to create a new EP
+        if err_msg == "404" or err_msg == "list index out of range":
+            create_new_ep(mac, attr_map, CISCO_ISE_ACTIVE_INSTANCE)
+
+        # 405 - Method not allowed means we need to switch to an old filter based API
+        elif err_msg == '405':
+            GET_EP_ID_CMD = "cisco-ise-get-endpoint-id"
+
+        # The primary went down (connection Error) or 401 if a fail over occurred (this primary/active
+        # is not a secondary/standby device).We should attempt to get the new Primary/Active
+        # instance is possible.
+        elif err_msg == "Connection Error" or err_msg == "401":
+            # Failover can take up to 10 minutes, its ok to just wait even if its a standalone ISE noe.
+            msg = "ISE instance is down. Trying again in 10 minutes. Error = %s" % err_msg
+            demisto.info("PANW_IOT_3RD_PARTY_BASE %s" % msg)
+            send_status_to_panw_iot_cloud("error", msg)
+            time.sleep(10 * 60)
```

Comment:
that's a lot of time to sleep. usually done in the PB level if needed. let's sync on that.